Addons/testExoticQuadrature.hpp

```diff
@@ -0,0 +1,162 @@
+    // n = 0, 1, 5
+    std::vector<int> n_vec = {1, 2, 5};
+    for (auto n : n_vec) {
+        roots = TasGrid::getRoots(0, ref_weights, ref_points);
+        if (roots.size() != 0) {
+            std::cout << "ERROR: Test failed in testRootSizes() for n = " << n << std::endl;
+            passed = false;
+        }
+    }
+    return passed;
+}
+
+// Test the attributes of the output of getExoticQuadrature().
+inline bool testBasicAttributes() {
+    bool passed = true;
+    const int n = 3;
+    const int nref = 11;
+    const char *descr = "TEST_DESCRIPTION";
+    auto sinc = [](double x) -> double {return (x == 0.0 ? 1.0 : sin(x) / (x));};
+    // Different behaviors expected for shift == 0.0 vs shift != 0.0.
+    std::vector<double> shifts = {0.0, 1.0};
+    for (auto shift : shifts) {
+        TasGrid::CustomTabulated ct = TasGrid::getExoticQuadrature(n, shift, sinc, nref, descr);
+        if (strcmp(ct.getDescription(), descr) != 0) {
+            std::cout << "ERROR: Test failed in testCtAttributes() for shift = " << shift << " on getDescription()"<< std::endl;
+            passed = false;
+        }
+        if (ct.getNumLevels() != n) {
+            std::cout << "ERROR: Test failed in testCtAttributes() for shift = " << shift << " on getNumLevels()" << std::endl;
+            passed = false;
+        }
+        for (int i=0; i<n; i++) {
+            if (ct.getQExact(i) != 2*(i+1)-1) {
+                std::cout << "ERROR: Test failed in testCtAttributes() for shift = " << shift << ", i = " << i
+                          << " on getQExact()" << std::endl;
+                passed = false;
+            }
+        }
+    }
+    return passed;
+}
+
+
+// Given a dimension, integrates the function f(x[1],...,x[dimension]) * sinc(freq * x[1]) * ... * sinc(freq * x[dimension]) over
+// [-1, 1]^dimension  using Exotic quadrature at a given depth and a specified shift. Then, compares this integral against a given exact
+// integral value.
+inline bool wrapSincTest(std::function<double(std::vector<double>)> f, int dimension, int depth, double freq, double shift,
+                         double exact_integral, double precision = 1e-10) {
+
+    // Initialize.
+    bool passed = true;
+    auto sinc = [freq](double x)->double {return (x == 0.0 ? 1.0 : sin(freq * x) / (freq * x));};
+    int level = depth % 2 == 1 ? (depth + 1) / 2 : depth / 2 + 1;
+    TasGrid::CustomTabulated ct = TasGrid::getExoticQuadrature(level, shift, sinc);
+    TasGrid::TasmanianSparseGrid sg;
+    sg.makeGlobalGrid(dimension, 1, depth, TasGrid::type_qptotal, std::move(ct));
+
+    // Compute the integral and compare to the reference value.
+    std::vector<double> quad_points = sg.getPoints();
+    std::vector<double> quad_weights = sg.getQuadratureWeights();
+    double approx_integral = 0.0;
+    assert(quad_weights.size() * dimension == quad_points.size());
+    for (size_t i=0; i<quad_weights.size(); i++) {
+        std::vector<double> x(dimension);
```

Comment:
If you need to copy, you should use the proper constructor:
```
std::vector<double> x(&quad_points[dimension * i], &quad_points[dimension * i] + dimension);
```

Reply:
Removed the need to copy.